File: Server/src/main/java/com/karandaev/roadCrackDetectionServer/image/validation/ImageByteValidator.java
package com.karandaev.roadCrackDetectionServer.image.validation;

import com.karandaev.roadCrackDetectionServer.image.config.ImageSecurityProperties;
import com.karandaev.roadCrackDetectionServer.image.exception.ImageValidationException;
import org.springframework.http.HttpStatus;

/** Validates raw uploaded image bytes before any expensive processing. */
public final class ImageByteValidator {

  private final ImageSecurityProperties props;

  /**
   * Creates an image byte validator.
   *
   * @param props image security configuration
   */
  public ImageByteValidator(ImageSecurityProperties props) {
    this.props = props;
  }

  /**
   * Validates that uploaded bytes are present and do not exceed the configured size limit.
   *
   * @param bytes raw uploaded file bytes
   * @throws ImageValidationException if the file is empty or too large
   */
  public void validateBytes(byte[] bytes) {
    if (bytes == null || bytes.length == 0) {
      throw new ImageValidationException(HttpStatus.BAD_REQUEST, "Empty file");
    }

    if (bytes.length > props.getMaxBytes()) {
      throw new ImageValidationException(HttpStatus.PAYLOAD_TOO_LARGE, "File too large");
    }
  }
}

File: MobileApp/app/src/main/java/com/karandaev/geo_inspect/core/image/validation/validation/ImageByteValidator.kt
package com.karandaev.geo_inspect.core.image.validation.validation

import com.karandaev.geo_inspect.core.image.validation.config.ImageSecurityProperties
import com.karandaev.geo_inspect.core.image.validation.exception.ImageValidationErrorCode
import com.karandaev.geo_inspect.core.image.validation.exception.ImageValidationException

/**
 * Validates raw uploaded image bytes before any expensive processing.
 *
 * @param props Image security configuration.
 */
class ImageByteValidator(
  private val props: ImageSecurityProperties
) {

  /**
   * Validates that uploaded bytes are present and do not exceed the configured size limit.
   *
   * @param bytes Raw uploaded file bytes.
   *
   * @throws ImageValidationException if the file is empty or too large.
   */
  fun validateBytes(bytes: ByteArray?) {
    if (bytes == null || bytes.isEmpty()) {
      throw ImageValidationException(
        code = ImageValidationErrorCode.EmptyFile,
        message = "Empty file."
      )
    }

    if (bytes.size > props.maxBytes) {
      throw ImageValidationException(
        code = ImageValidationErrorCode.FileTooLarge,
        message = "File is too large."
      )
    }
  }
}